File: qqbot-module-api/src/main/java/com/mieai/qqbot/module/api/ModuleDependency.java
package com.mieai.qqbot.module.api;

/** A versioned dependency on another framework module. */
public record ModuleDependency(String moduleId, String minimumVersion, boolean optional) {
    public ModuleDependency {
        moduleId = ModuleValidation.requireId(moduleId, "moduleId");
        minimumVersion = ModuleValidation.requireVersion(minimumVersion, "minimumVersion");
    }

    public static ModuleDependency required(String moduleId, String minimumVersion) {
        return new ModuleDependency(moduleId, minimumVersion, false);
    }

    public static ModuleDependency optional(String moduleId, String minimumVersion) {
        return new ModuleDependency(moduleId, minimumVersion, true);
    }
}

File: qqbot-module-api/src/main/kotlin/com/mieai/qqbot/module/api/ModuleDependency.kt
package com.mieai.qqbot.module.api


/** A versioned dependency on another framework module. */
data class ModuleDependency(
    val moduleId: String,
    val minimumVersion: String,
    val optional: Boolean,
) {
    init {
        ModuleValidation.requireId(moduleId, "moduleId")
        ModuleValidation.requireVersion(minimumVersion, "minimumVersion")
    }

    companion object {
        fun required(moduleId: String, minimumVersion: String): ModuleDependency =
            ModuleDependency(moduleId, minimumVersion, false)

        fun optional(moduleId: String, minimumVersion: String): ModuleDependency =
            ModuleDependency(moduleId, minimumVersion, true)
    }
}
